engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/RuntimeCache.java

```diff
@@ -1,17 +1,22 @@
 package org.enso.interpreter.instrument;
 
 import com.oracle.truffle.api.CompilerDirectives;
+import java.lang.ref.Reference;
 import java.lang.ref.SoftReference;
+import java.lang.ref.WeakReference;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Set;
 import java.util.UUID;
+import java.util.WeakHashMap;
 import org.enso.interpreter.service.ExecutionService;
 
 /** A storage for computed values. */
-public final class RuntimeCache {
+public final class RuntimeCache implements java.util.function.Function<String, Object> {
 
-  private final Map<UUID, SoftReference<Object>> cache = new HashMap<>();
+  private final Map<UUID, Reference<Object>> cache = new HashMap<>();
+  private final Map<Object, UUID> valuesToKeys = new WeakHashMap<>();
+  private final Map<UUID, Reference<Object>> expressions = new HashMap<>();
```

Comment:
I'm concerned that these new caches are not invalidated.

Reply:
The invalidation is happening in the compile job https://github.com/enso-org/enso/blob/d665f4d9c22088c3be8fbbffd7224e8a681b5926/engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/EnsureCompiledJob.scala#L372-L383

Reply:
Yes, these two new caches aren't invalidated. However that shouldn't matter much - both are using `WeakReference` and as such they don't hold the referenced objects in memory more than `cache` which uses `SoftReference` - e.g. it is necessary to clean the `cache` only, the `valuesToKeys` and `expressions` are going to be cleaned occasionally then by GC.

Btw. `valuesToKeys` is a `WeakHashMap` - e.g. the keys are hold by `WeakReference` automatically. `expressions` is using `WeakReference` explicitly to hold the values.

I can write some tests, if necessary to prove the system is behaving correctly. Other than that there shall be no need to change the code.

Reply:
Test written in 63c0b20173 - we don't have an `assertGC` test in the code base yet, we'll see how flaky it'll be. However the [assertGC concept](https://github.com/apache/netbeans/blob/f01e76e49ac5996e0e41589e722ad9ccb508144f/harness/nbjunit/src/org/netbeans/junit/NbTestCase.java#L1434) has been sufficiently used and tested already.